examples/multiple_instances/reader.py

```diff
@@ -0,0 +1,76 @@
+"""A simple bot (named benchmark-reader), reading a datafeed and replying to a message with the same content.
+The goal is to run multiple instances of it, hence it adds a unique id and uses a random color in the reply.
+"""
+import asyncio
+import logging.config
+import uuid
+from pathlib import Path
+from random import randrange
+
+import hazelcast
+from hazelcast import HazelcastClient
+
+from symphony.bdk.core.config.loader import BdkConfigLoader
+from symphony.bdk.core.service.datafeed.real_time_event_listener import RealTimeEventListener
+from symphony.bdk.core.service.message.message_parser import get_text_content_from_message
+from symphony.bdk.core.service.message.message_service import MessageService
+from symphony.bdk.core.symphony_bdk import SymphonyBdk
+from symphony.bdk.gen.agent_model.v4_initiator import V4Initiator
+from symphony.bdk.gen.agent_model.v4_message_sent import V4MessageSent
+
+
+class EventListener(RealTimeEventListener):
+
+    def __init__(self, messages: MessageService, hz_client: HazelcastClient):
+        self._messages = messages
+        self._client = hz_client
+        self._id = str(uuid.uuid4())
+        self._color = hex(randrange(0xffffff + 1))[2:]  # substring to remove leading "0x"
+
+    async def on_message_sent(self, initiator: V4Initiator, event: V4MessageSent):
+        lock = self._client.cp_subsystem.get_lock(event.message.stream.stream_id).blocking()
+        lock.lock()
+        try:
+            return await self._reply_to_message(event)
+        finally:
+            lock.unlock()
+
+    async def _reply_to_message(self, event):
+        message = event.message
+        text_message = get_text_content_from_message(message)
+        message_number = text_message.split(" ")[0]
+        logging.debug("Received message number %s", message_number)
+
+        processed_events = self._client.get_map("processed_events").blocking()
```

Comment:
can it be a problem to block from an async function?

Reply:
Indeed it is blocking but asynchronous maps in the hazelcast lib does not used asyncio but callbacks (see https://hazelcast.readthedocs.io/en/stable/api/proxy/map.html). This would make the implementation harder and less readable IMO.